src/librustc_typeck/collect.rs

```diff
@@ -2256,6 +2256,95 @@ fn predicates_from_bound<'tcx>(
     }
 }
 
+/// Returns the minimum required target-feature name to use a SIMD type on C FFI:
+fn simd_ffi_min_target_feature(simd_width: usize) -> Option<&'static str> {
+    // FIXME: this needs to be architecture dependent and
+    // should probably belong somewhere else:
+    // * on arm: 8 => neon
+    // * on aarch64: 8 | 16 => neon,
+    // * on mips: 16 => msa,
+    // * on ppc: 16
+    //     if vec elem is 64-bit wide => vsx
+    //     otherwise => altivec
+    // * wasm: 16 => simd128
+    match simd_width {
+        16 => Some("sse"),
+        32 => Some("avx"),
+        64 => Some("avx512"),
+        _ => None,
+    }
+}
+
+/// Returns true if the target-feature allows using the SIMD type on C FFI:
+fn simd_ffi_feature_check(simd_width: usize, feature: &'static str) -> bool {
+    // FIXME: see simd_ffi_min_target_feature
+    match simd_width {
+        16 => feature.contains("sse") || feature.contains("ssse") || feature.contains("avx"),
+        32 => feature.contains("avx"),
+        64 => feature.contains("avx512"),
+        _ => false,
+    }
+}
+
+fn simd_ffi_check<'a, 'tcx, 'b: 'tcx>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId, ast_ty: &hir::Ty, ty: Ty<'b>,
+    attrs: &mut Option<CodegenFnAttrs>
+) {
+    if !ty.is_simd() {
+        return;
+    }
+
+    // The use of SIMD types in FFI is feature-gated:
+    if !tcx.features().simd_ffi {
+        tcx.sess
+            .struct_span_err(
+                ast_ty.span,
+                &format!(
+                    "use of SIMD type `{}` in FFI is unstable",
+                    tcx.hir().hir_to_pretty_string(ast_ty.hir_id)
+                ),
+            )
+            .help("add #![feature(simd_ffi)] to the crate attributes to enable")
+            .emit();
+        return;
+    }
+
+    if attrs.is_none() {
+        *attrs = Some(codegen_fn_attrs(tcx, def_id));
```

Comment:
I think the cycle may be happening here?

Instead of memoizing `codegen_fn_attrs` across function calls you'll want to use `tcx.codegen_fn_attrs(def_id)` which does memoization for you as well. Somewhere along the way of the implementation of `codegen_fn_attrs`, though, it may require the signature for the type which then requires this, which is where the whole cycle may be. In the end you may have to split out the SIMD check to happen in a different code path that isn't transitively called from the `codegen_fn_attrs` query

Reply:
Ah! That makes sense, I'll give that a try!

Reply:
Ok so the problem here is that `codegen_fn_attrs` type checks `#[target_feature]` by making sure that the `fn` is unsafe, by callign `fn_sig.unsafe()` which calls the foreign type checking function for FFI functions.

Reply:
I feel that `codegen_fn_attrs` should just return the function attributes, and `fn_sig` should query them and if the function has `#[target_feature]` then it should make sure that it is `unsafe`.